[Bundle 3/n] Prune remaining g_chainman usage in validation functions

dongcarl opened the pull request:
Overall PR: #20158 (tree-wide: De-globalize ChainstateManager)

Based on:
- [x] #20750 | [Bundle 2/n] Prune g_chainman usage in mempool-related validation functions

Note to reviewers:
1. This bundle may _apparently_ introduce usage of `g_chainman` or `::Chain(state|)Active()` globals, but these are resolved later on in the overall PR. [Commits of overall PR](https://github.com/bitcoin/bitcoin/pull/20158/commits)
2. There may be seemingly obvious local references to `ChainstateManager` or other validation objects which are not being used in callers of the current function in question, this is done intentionally to **_keep each commit centered around one function/method_** to ease review and to make the overall change systematic. We don't assume anything about our callers. Rest assured that once we are considering that particular caller in later commits, we will use the obvious local references. [Commits of overall PR](https://github.com/bitcoin/bitcoin/pull/20158/commits)
3. When changing a function/method that has many callers (e.g. `LookupBlockIndex` with 55 callers), it is sometimes easier (and less error-prone) to use a scripted-diff. When doing so, there will be 3 commits in sequence so that every commit compiles like so:
	1. Add `new_function`, make `old_function` a wrapper of `new_function`, divert all calls to `old_function` to `new_function` **in the local module only**
	2. Scripted-diff to divert all calls to `old_function` to `new_function` **in the rest of the codebase**
	3. Remove `old_function`

Note to self:
- [x] Address: https://github.com/bitcoin/bitcoin/pull/20750#discussion_r579400663

dongcarl committed:
validation: Remove global ::LoadGenesisBlock

dongcarl committed:
validation: Move LoadBlockIndexDB to CChainState

CChainState needed cuz setBlockIndexCandidates

dongcarl committed:
validation: Pass in chainstate to CTxMemPool::check

This is the only instance where validation reaches for something outside
of it.

dongcarl committed:
validation: Remove global ::{{Precious,Invalidate}Block,ResetBlockFailureFlags}

dongcarl committed:
validation: Pass in chainstate to UpdateTip

dongcarl committed:
validation: Pass in chainstate to ::PruneBlockFilesManual

dongcarl committed:
validation: Remove global ::VersionBitsTip{State,SinceHeight,Statistics}

Tip: versionbitscache is currently a global so we didn't need to pass it
     in to any of ::VersionBitsTip*'s callers

dongcarl committed:
validation: Pass in chainstate to CVerifyDB::VerifyDB

dongcarl committed:
validation: Use *this in CChainState::ActivateBestChainStep

dongcarl committed:
validation: Use *this in CChainState::InvalidateBlock

dongcarl committed:
validation: Move invalid block handling to CChainState

- InvalidChainFound
- CheckForkWarningConditions

dongcarl committed:
validation: Pass in chain to FindBlockPos+SaveBlockToDisk

dongcarl committed:
validation: Use existing chain member in CChainState::AcceptBlock

dongcarl committed:
validation: Use existing chain member in CChainState::LoadGenesisBlock

ryanofsky commented on src/validation.cpp:
```diff
@@ -4112,11 +4112,12 @@ void BlockManager::Unload() {
 }
 
-bool static LoadBlockIndexDB(ChainstateManager& chainman, const CChainParams& chainparams) EXCLUSIVE_LOCKS_REQUIRED(cs_main)
+bool CChainState::LoadBlockIndexDB(const CChainParams& chainparams) EXCLUSIVE_LOCKS_REQUIRED(cs_main)
```
In commit "validation: Move LoadBlockIndexDB to CChainState" (fa730f7a871c49072e2e899cefbc8275e5f67057)

It seems chainman variable wasn't really needed here, just the chainman.m_blockman member, so this might make more sense as a BlockManager method instead of CChainState method

dongcarl replied:
Unfortunately we need `setBlockIndexCandidates` which is a member off `CChainState`... If you think `LoadBlockIndexDB` makes more sense logically as a method of `BlockManager` rather than `CChainState` we can pass in the `setBlockIndexCandidates`. Let me know!

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/21055#discussion_r578720086

> Unfortunately we need `setBlockIndexCandidates` which is a member off `CChainState`... If you think `LoadBlockIndexDB` makes more sense logically as a method of `BlockManager` rather than `CChainState` we can pass in the `setBlockIndexCandidates`. Let me know!

Thanks, I didn't recognize the `setBlockIndexCandidates` access. Proposed change sounds reasonable, but I don't actually know too much about it and it would seem good to tackle as a followup not changing here.

ryanofsky commented on src/net_processing.cpp:
```diff
@@ -3201,7 +3203,10 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
-            m_mempool.check(&::ChainstateActive().CoinsTip());
+            CChainState& active_chainstate = m_chainman.ActiveChainstate();
+            CCoinsViewCache& active_coins_tip = active_chainstate.CoinsTip();
+            m_mempool.check(&active_coins_tip, active_chainstate.m_blockman.GetSpendHeight(active_coins_tip));
```
In commit "validation: Pass in spendheight to CTxMemPool::check" (a765747415aebd432d5feffb6cb609be62e2592b)

I think it would be nice to delete the BlockManager::GetSpendHeight method entirely instead of adding new calls to it. It's basically a one-line function returning height of active_coins_tip + 1, and it would seem more straightforward to pass active_coins_tip and active_coins_tip height together as arguments to the mempool check method (assuming the check method can't access the height directly) instead of the tip and an offset height.

dongcarl replied:
Not entirely sure what you mean here by an "offset height"... Looking at all the uses of `spendheight` in `CTxMemPool::check` it seems that this "offset height" is exactly what `CheckInputsAndUpdateCoins` wants?

jnewbery replied:
I agree with @ryanofsky that the use of `BlockManager::GetSpendHeight()` is quite ugly. We can get the height directly from `::ChainstateActive().m_chain.Height() + 1`.

I also think it'd be preferable to not have net_processing reach inside validation for the various active chainstate fields, just to call a `check()` function (which defaults to a no-op with default `-mempoolcheck`). Could we perhaps move this call to `CTxMemPool.check()` into `AcceptToMemoryPool()`, so that it's done consistently for all callers, and net_processing doesn't need to concern itself with validation internals?

dongcarl replied:
I'm not sure I can easily convince myself that moving this check into `ATMP` is completely safe. However, I do see how this is ugly.

Do you think the following patch would be an acceptable solution?

```diff
diff --git a/src/net_processing.cpp b/src/net_processing.cpp
index 614a3e0791..022f2e0574 100644
--- a/src/net_processing.cpp
+++ b/src/net_processing.cpp
@@ -2301,10 +2301,7 @@ void PeerManagerImpl::ProcessOrphanTx(std::set<uint256>& orphan_work_set)
             break;
         }
     }
-    CChainState& active_chainstate = m_chainman.ActiveChainstate();
-    CCoinsViewCache& active_coins_tip = active_chainstate.CoinsTip();
-    assert(std::addressof(::ChainstateActive().CoinsTip()) == std::addressof(active_coins_tip));
-    m_mempool.check(&active_coins_tip, active_chainstate.m_blockman.GetSpendHeight(active_coins_tip));
+    m_mempool.check(m_chainman.ActiveChainstate());
 }
 
 /**
@@ -3265,10 +3262,7 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
         const TxValidationState& state = result.m_state;
 
         if (result.m_result_type == MempoolAcceptResult::ResultType::VALID) {
-            CChainState& active_chainstate = m_chainman.ActiveChainstate();
-            CCoinsViewCache& active_coins_tip = active_chainstate.CoinsTip();
-            assert(std::addressof(::ChainstateActive().CoinsTip()) == std::addressof(active_coins_tip));
-            m_mempool.check(&active_coins_tip, active_chainstate.m_blockman.GetSpendHeight(active_coins_tip));
+            m_mempool.check(m_chainman.ActiveChainstate());
 
             // As this version of the transaction was acceptable, we can forget about any
             // requests for it.
diff --git a/src/txmempool.cpp b/src/txmempool.cpp
index 4df7017c47..05902a55c3 100644
--- a/src/txmempool.cpp
+++ b/src/txmempool.cpp
@@ -619,7 +619,7 @@ static void CheckInputsAndUpdateCoins(const CTransaction& tx, CCoinsViewCache& m
     UpdateCoins(tx, mempoolDuplicate, std::numeric_limits<int>::max());
 }
 
-void CTxMemPool::check(const CCoinsViewCache *pcoins, const int64_t spendheight) const
+void CTxMemPool::check(CChainState& active_chainstate) const
 {
     if (m_check_ratio == 0) return;
 
@@ -633,7 +633,10 @@ void CTxMemPool::check(const CCoinsViewCache *pcoins, const int64_t spendheight)
     CAmount check_total_fee{0};
     uint64_t innerUsage = 0;
 
-    CCoinsViewCache mempoolDuplicate(const_cast<CCoinsViewCache*>(pcoins));
+    CCoinsViewCache& active_coins_tip = active_chainstate.CoinsTip();
+    assert(std::addressof(::ChainstateActive().CoinsTip()) == std::addressof(active_coins_tip)); // TODO: REVIEW-ONLY, REMOVE IN FUTURE COMMIT
+    CCoinsViewCache mempoolDuplicate(const_cast<CCoinsViewCache*>(&active_coins_tip));
+    const int64_t spendheight = active_chainstate.m_chain.Height() + 1;
     assert(g_chainman.m_blockman.GetSpendHeight(mempoolDuplicate) == spendheight); // TODO: REVIEW-ONLY, REMOVE IN FUTURE COMMIT
 
     std::list<const CTxMemPoolEntry*> waitingOnDependants;
@@ -655,7 +658,7 @@ void CTxMemPool::check(const CCoinsViewCache *pcoins, const int64_t spendheight)
                 fDependsWait = true;
                 setParentCheck.insert(*it2);
             } else {
-                assert(pcoins->HaveCoin(txin.prevout));
+                assert(active_coins_tip.HaveCoin(txin.prevout));
             }
             // Check whether its inputs are marked in mapNextTx.
             auto it3 = mapNextTx.find(txin.prevout);
diff --git a/src/txmempool.h b/src/txmempool.h
index dd82d61fc6..001d856e43 100644
--- a/src/txmempool.h
+++ b/src/txmempool.h
@@ -604,7 +604,7 @@ public:
      * all inputs are in the mapNextTx array). If sanity-checking is turned off,
      * check does nothing.
      */
-    void check(const CCoinsViewCache *pcoins, const int64_t spendheight) const EXCLUSIVE_LOCKS_REQUIRED(::cs_main);
+    void check(CChainState& active_chainstate) const EXCLUSIVE_LOCKS_REQUIRED(::cs_main);
 
     // addUnchecked must updated state for all ancestors of a given transaction,
     // to track size/count of descendant transactions.  First version of
diff --git a/src/validation.cpp b/src/validation.cpp
index 325edc5b6a..ac79dce52e 100644
--- a/src/validation.cpp
+++ b/src/validation.cpp
@@ -2801,8 +2801,7 @@ bool CChainState::ActivateBestChainStep(BlockValidationState& state, const CChai
         UpdateMempoolForReorg(*this, m_mempool, disconnectpool, true);
     }
 
-    CCoinsViewCache& active_coins_tip = CoinsTip();
-    m_mempool.check(&active_coins_tip, m_blockman.GetSpendHeight(active_coins_tip));
+    m_mempool.check(*this);
 
     CheckForkWarningConditions();
 
```

jnewbery replied:
> I'm not sure I can easily convince myself that moving this check into ATMP is completely safe.

I think you're right to be wary. The call to ATMP from UpdateMempoolForReorg probably can't call `check()`.

I like your proposed change.

dongcarl replied:
Fixed in 4744efc9ba

ryanofsky reviewed (approved):
Code review ACK f5ee2f37742da70d1731508e647dccd9b6035a7c

ariard commented on src/validation.cpp:
```diff
@@ -4973,24 +4973,6 @@ CBlockFileInfo* GetBlockFileInfo(size_t n)
 
-ThresholdState VersionBitsTipState(const Consensus::Params& params, Consensus::DeploymentPos pos)
-{
-    LOCK(cs_main);
```
This global was introduced by d23f6c6. At that commit, the global`VersionBitsTipState`'s unique caller`BIP9SoftForkDesc` didn't require `cs_main` locking. 

Those methods were refactored by 3862e47, where the new `BIP9SoftForkDescPushBack` did require the lock. I think since then this lock has been useless and could have been substituted by an annotation.

So not a concern to get rid of them.

ariard commented on src/net_processing.cpp:
```diff
@@ -2243,7 +2243,9 @@ void PeerManagerImpl::ProcessOrphanTx(std::set<uint256>& orphan_work_set)
-    m_mempool.check(&::ChainstateActive().CoinsTip());
+    CChainState& active_chainstate = m_chainman.ActiveChainstate();
+    CCoinsViewCache& active_coins_tip = active_chainstate.CoinsTip();
+    m_mempool.check(&active_coins_tip, active_chainstate.m_blockman.GetSpendHeight(active_coins_tip));
```
Shouldn't an assert be added here and before the following invocations ? The `CCoinsViewCache` pointer is used throughout `CTxMemPool::check.

dongcarl replied:
Fixed in: 52c29d9aedcf8d141b75a6d923a5d5bea53f4079

ariard reviewed (commented):
See about the maybe missing assert otherwise code review ACK f5ee2f3

dongcarl commented:
Pushed f5ee2f37742da70d1731508e647dccd9b6035a7c -> eb8f8e42dcca3448ceddb2d774342a524475dd6e
- Rebased on master

> Note: Have not yet addressed review.

dongcarl commented:
Pushed eb8f8e42dcca3448ceddb2d774342a524475dd6e -> 914e9d321608b909f7f8f0442fa64cb425fcd955
- Addressed: https://github.com/bitcoin/bitcoin/pull/21055#discussion_r572046550

jnewbery commented:
Concept ACK. This may need rebase now that #20750 is merged.

dongcarl commented:
@MarcoFalke w/re https://github.com/bitcoin/bitcoin/pull/20750#discussion_r579400663, it seems like making the change would require changing all callers as well (sorta out of scope for this PR), how about we just do the following for now to ensure that the theoretical UB won't happen?

```diff
diff --git a/src/validation.cpp b/src/validation.cpp
index a66c3d09e3..2bd7046e0f 100644
--- a/src/validation.cpp
+++ b/src/validation.cpp
@@ -211,6 +211,7 @@ static FlatFileSeq UndoFileSeq();
 bool CheckFinalTx(const CBlockIndex* active_chain_tip, const CTransaction &tx, int flags)
 {
     AssertLockHeld(cs_main);
+    assert(active_chain_tip); // TODO: Make active_chain_tip a reference
     assert(std::addressof(*::ChainActive().Tip()) == std::addressof(*active_chain_tip));
 
     // By convention a negative value for flags indicates that the
```

dongcarl commented:
Pushed 914e9d321608b909f7f8f0442fa64cb425fcd955 -> 1e0d2d83f6ba8761eee4d2b97828e18204804dc3
- Rebased on master

dongcarl committed:
validation: Check chain tip is non-null in CheckFinalTx

...also update comments to remove mention of ::ChainActive()

From: https://github.com/bitcoin/bitcoin/pull/20750#discussion_r579400663

> Also, what about passing a const reference instead of a pointer? I
> know this is only theoretical, but previously if the tip was nullptr,
> then Height() evaluated to -1, now it evaluates to UB

dongcarl commented:
Pushed 1e0d2d83f6ba8761eee4d2b97828e18204804dc3 → 0bd83b71afd7cd60bca6870ce9428b9fe5bc120e
- Prepended commit 9da106be4db692fa5db7b4de79f9cf7bfef37075 to address https://github.com/bitcoin/bitcoin/pull/20750#discussion_r579400663

dongcarl committed:
validation: CVerifyDB::VerifyDB: Use locking annotation

...instead of recursively locking unconditionally

dongcarl commented:
Pushed 0bd83b71afd7cd60bca6870ce9428b9fe5bc120e -> ef83bb85507ad9e4e6801aa2197f195e9e42a948
- Addressed some of jnewbery's review comments

dongcarl commented on src/validation.cpp:
```diff
@@ -2445,7 +2449,9 @@ static void UpdateTip(CTxMemPool& mempool, const CBlockIndex* pindexNew, const C
-    if (!::ChainstateActive().IsInitialBlockDownload()) {
+    assert(std::addressof(::ChainstateActive()) == std::addressof(active_chainstate));
+    if (!active_chainstate.IsInitialBlockDownload())
+    {
```
Fixed in 4bada76237

dongcarl commented on src/validation.cpp:
```diff
@@ -4242,15 +4246,16 @@ CVerifyDB::~CVerifyDB()
 }
 
-bool CVerifyDB::VerifyDB(const CChainParams& chainparams, CCoinsView *coinsview, int nCheckLevel, int nCheckDepth)
+bool CVerifyDB::VerifyDB(const CChainParams& chainparams, CChainState& active_chainstate, CCoinsView *coinsview, int nCheckLevel, int nCheckDepth)
```
Fixed in ef83bb8550

jnewbery replied:
Perhaps also add `AssertLockHeld(cs_main)` to the top of the function body?

dongcarl replied:
Fixed in e11b649650

dongcarl commented on src/validation.h:
```diff
@@ -776,21 +773,14 @@ class CChainState
- */
-bool PreciousBlock(BlockValidationState& state, const CChainParams& params, CBlockIndex *pindex) LOCKS_EXCLUDED(cs_main);
+    void CheckForkWarningConditions() EXCLUSIVE_LOCKS_REQUIRED(cs_main);
+    void CheckForkWarningConditionsOnNewFork(CBlockIndex* pindexNewForkTip) EXCLUSIVE_LOCKS_REQUIRED(cs_main);
```
Fixed in: 8b99efbcc0

dongcarl commented on src/validation.cpp:
```diff
@@ -4140,11 +4143,12 @@ void BlockManager::Unload() {
 }
 
-bool static LoadBlockIndexDB(ChainstateManager& chainman, const CChainParams& chainparams) EXCLUSIVE_LOCKS_REQUIRED(cs_main)
+bool CChainState::LoadBlockIndexDB(const CChainParams& chainparams) EXCLUSIVE_LOCKS_REQUIRED(cs_main)
```
Fixed in 8cdb2f7e58

jnewbery commented:
Changes look good, @dongcarl! I've left a couple more comments.

dongcarl commented:
Pushed ef83bb85507ad9e4e6801aa2197f195e9e42a948 -> e11b6496506246882df450586acf735dabedf731
- Applied https://github.com/bitcoin/bitcoin/pull/21055#discussion_r585113995
- Addressed https://github.com/bitcoin/bitcoin/pull/21055#discussion_r585535526

Thanks jnewbery for your diligent review!

laanwj commented:
Code review ACK e11b6496506246882df450586acf735dabedf731

ryanofsky reviewed (commented):
Code review ACK e11b6496506246882df450586acf735dabedf731. Changes since last review: rebase, two new commits adding null assert and lock assert, replacing mempool check coins and spendheight arguments
